### PR TITLE
feat: timeout jobs

### PR DESCRIPTION
**Pull request by ZeRego**
Closes: #5123 

### Description:

Context: 
We can reject a job affect X seconds but it is not possible to cancel a promise.
eg: we can mark a job as failed after 1 minute but since the promise is not canceled it will still send the email after 2 minutes.

Code in this PR was based in the example in this blog post: https://advancedweb.hu/how-to-add-timeout-to-a-promise-in-javascript/

Testing using a 1 minute timeout, was able to download a csv:

Captura de ecrã 2023-04-28, às 16 54 20

Testing using a 1 minute timeout, and force the csv download to take at least 2 minutes. We can see the job timed out and is marked as completed to prevent any retries.

Captura de ecrã 2023-04-28, às 16 57 19

After 2 minutes the csv download terminated.

Captura de ecrã 2023-04-28, às 16 58 44

Schedule history shows error:
Captura de ecrã 2023-05-02, às 15 12 50

**TLDR:** 

We now mark jobs as completed if they take more than 10 minutes to finish.

### CODE REVIEW

File: packages/backend/src/scheduler/SchedulerJobTimeout.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    let timer: NodeJS.Timeout;
+    return Promise.race([
+        prom,
+        new Promise((_, reject) => {
+            timer = setTimeout(reject, time, exception);
+        }),
+    ]).finally(() => clearTimeout(timer));
+}
+
+const DEFAULT_JOB_TIMEOUT = 1000 * 60 * 10; // 10 minutes
```
Comment (rephus) [inline packages/backend/src/scheduler/SchedulerJobTimeout.ts:14]: We might want to create an env for this. But I'm ok as it is

File: packages/backend/src/scheduler/SchedulerJobTimeout.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    try {
+        await timeout(prom, time, timeoutError);
+    } catch (e) {
+        if (e !== timeoutError) {
+            // propagate non-timeout errors
+            throw e;
+        } else {
+            // handle timeout
+            Logger.error(
+                `Worker ${job.locked_by} timed out job ${job.id} (${job.task_identifier}) after ${time}ms`,
```
Comment (rephus) [inline packages/backend/src/scheduler/SchedulerJobTimeout.ts:30]: If a job timesout, it will appear as In progress in our scheduler logs, right? Should we also create an error entry here?

**Review by rephus — APPROVED** [review]
Just added a question, but I'm happy to merge this as it is

File: packages/backend/src/scheduler/SchedulerTask.ts [hunk 3]
```diff
@@ -538,7 +538,7 @@ export const sendEmailNotification = async (
             scheduledTime,
             targetType: 'email',
             status: SchedulerJobStatus.ERROR,
-            details: e,
+            details: { error: e.message },
```
Comment (ZeRego) [inline packages/backend/src/scheduler/SchedulerTask.ts:541]: Fixed this to have the same details structure as the other logs.

File: packages/backend/src/scheduler/SchedulerJobTimeout.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            );
+            const timeOutError = new Error(`Job timed out after ${time}ms`);
+            Sentry.captureException(timeOutError, {
+                extra: {
+                    jobId: job.id,
+                    workerId: job.locked_by,
+                    task: job.task_identifier,
+                },
+            });
+            await onTimeout(job, timeOutError);
```
Comment (rephus) [inline packages/backend/src/scheduler/SchedulerJobTimeout.ts:41]: Nice

**Review by rephus — APPROVED** [review]
Perfection
